File: frontend/src/api/mcp.ts
import { apiClient, ApiResponse } from './client';

export type MCPTransport = 'stdio' | 'sse' | 'streamable-http';

export interface MCPServerInfo {
  name: string;
  transport: MCPTransport;
  enabled: boolean;
  description?: string;
  command?: string;
  args?: string[];
  url?: string;
  headers?: Record<string, string>;
  env?: Record<string, string>;
  scope?: 'global' | 'user';
  user_id?: string | null;
  owner_user_id?: string | null;
  installed?: boolean;
  source?: 'official' | 'personal' | 'community';
}

export interface MCPServerListResponse {
  servers: MCPServerInfo[];
}

export interface MCPServerUpsertResponse {
  server: MCPServerInfo;
}

export async function listMCPServers(): Promise<MCPServerInfo[]> {
  const response = await apiClient.get<ApiResponse<MCPServerListResponse>>('/mcp/servers');
  return response.data.data.servers;
}

export async function listMCPCatalog(): Promise<MCPServerInfo[]> {
  const response = await apiClient.get<ApiResponse<MCPServerListResponse>>('/mcp/servers/catalog');
  return response.data.data.servers;
}

export async function installMCPServer(name: string): Promise<MCPServerInfo> {
  const response = await apiClient.post<ApiResponse<MCPServerInfo>>(`/mcp/servers/${encodeURIComponent(name)}/install`);
  return response.data.data;
}

export async function uninstallMCPServer(name: string): Promise<MCPServerInfo> {
  const response = await apiClient.delete<ApiResponse<MCPServerInfo>>(`/mcp/servers/${encodeURIComponent(name)}/install`);
  return response.data.data;
}

export async function saveMCPServer(server: MCPServerInfo): Promise<MCPServerInfo> {
  const response = await apiClient.put<ApiResponse<MCPServerUpsertResponse>>(
    `/mcp/servers/${encodeURIComponent(server.name)}`,
    server,
  );
  return response.data.data.server;
}

export async function deleteMCPServer(name: string): Promise<MCPServerInfo[]> {
  const response = await apiClient.delete<ApiResponse<MCPServerListResponse>>(
    `/mcp/servers/${encodeURIComponent(name)}`,
  );
  return response.data.data.servers;
}

File: backend/app/interfaces/schemas/mcp.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel

from app.domain.models.mcp_config import MCPRiskLevel, MCPScope, MCPTransport


class MCPServerResponse(BaseModel):
    name: str
    transport: MCPTransport
    enabled: bool
    description: Optional[str] = None
    command: Optional[str] = None
    args: Optional[List[str]] = None
    url: Optional[str] = None
    headers: Optional[Dict[str, str]] = None
    env: Optional[Dict[str, str]] = None
    scope: MCPScope = MCPScope.GLOBAL
    user_id: Optional[str] = None
    owner_user_id: Optional[str] = None
    workspace_id: Optional[str] = None
    risk_level: MCPRiskLevel = MCPRiskLevel.STANDARD
    installed: bool = False
    source: str = "personal"


class MCPServerListResponse(BaseModel):
    servers: List[MCPServerResponse]


class MCPServerRequest(BaseModel):
    name: str
    transport: MCPTransport
    enabled: bool = True
    description: Optional[str] = None
    command: Optional[str] = None
    args: Optional[List[str]] = None
    url: Optional[str] = None
    headers: Optional[Dict[str, str]] = None
    env: Optional[Dict[str, str]] = None
    is_global: bool = False
    risk_level: MCPRiskLevel = MCPRiskLevel.STANDARD


class MCPServerUpsertResponse(BaseModel):
    server: MCPServerResponse


class MCPToolResponse(BaseModel):
    name: str
    server: str
    description: str
    parameters: Dict[str, Any]


class MCPToolListResponse(BaseModel):
    tools: List[MCPToolResponse]


class MCPToolListRequest(BaseModel):
    selected_servers: List[str] = []
